Yoast/wordpress-seo
# Improve user input validation for URL fields

## Summary

This PR can be summarized in the following changelog entry:

* Improved user input validation for URLs

This PR only improves the existing `WPSEO_Option->validate_url()` method. A proper user input validation should include better error reporting, clear identification of the input with the error, etc. These parts will be addressed in future PRs.

## Relevant technical choices:
`WPSEO_Option->validate_url()` is supposed to render a WP error notice when an invalid URL is submitted. However, its logic fails (see https://github.com/Yoast/bugreports/issues/457) and a "success" notice is always displayed, regardless of the URL is valid or invalid:

<img width="646" alt="screenshot 2019-03-06 at 11 17 13" src="https://user-images.githubusercontent.com/1682452/53874710-49474d80-4003-11e9-97bc-65e90fcea188.png">

Seems to me a better approach for _user input validation_ is to:
- first: validate the user input checking if the entered value matches the required data type, in this case: a valid URL
- then sanitize it

For this purpose, I'm using `filter_var()` and `FILTER_VALIDATE_URL`

## Test instructions

`WPSEO_Option->validate_url()` is currently used at these locations:

**SEO > Social:**
case 'facebook_site':
case 'instagram_url':
case 'linkedin_url':
case 'myspace_url':
case 'pinterest_url':
case 'og_default_image':
case 'og_frontpage_image':
case 'youtube_url':
case 'wikipedia_url':

**SEO > Search Appearance > General:**
case 'company_logo':
case 'person_logo':

Test all these cases, first on trunk:
- enter an invalid value e.g. `q w e r t y` with spaces
- see it gets:
  - prepended with `http://`
  - sanitized: the spaces become `%20`
  - as such, it's considered a "valid" URL so there's no error notice
  - instead, a "success" notice appears

Switch to this branch
- repeat the steps above
- see that `q w e r t y` gets first validated to check if it's a valid URL
- an error notice appears
- the URL field stays empty
- then enter a valid URL
- see it gets saved correctly
- then enter again an invalid URL
- an error notice appears
- the input field is filled with the previous valid value

Note: for the image fields:

case 'og_default_image':
case 'og_frontpage_image':
case 'company_logo':
case 'person_logo':

the UI actually doesn't allow to manually enter a value. These fields are readonly. The UI only allows to select an image from the WP Media Library so the URL is always valid (hopefully) and validation is a bit redundant, though it doesn't harm.

Note to test `og_frontpage_image` you have to have your homepage set to your latest posts. See: https://kb.yoast.com/kb/custom-open-graph-tags/

Adjusted 2 related tests that used an _absolute URL path_ e.g. `/test.png` for the og_default_image and the frontpage_image. These aren't valid URLs, and the UI doesn't allow to use them anyways.

* Because we've had some issues with the multisite tests, please also test on multisite.

## UI changes

## Documentation

Fixes https://github.com/Yoast/bugreports/issues/457
